api/src/main/java/jakarta/enterprise/invoke/InvokerBuilder.java

```diff
@@ -0,0 +1,375 @@
+ * where {@code X}, {@code Y} and {@code Z} are types:
+ *
+ * <ul>
+ * <li>{@code static Z wrap(X instance, Object[] arguments, Invoker<X, Y> invoker)}</li>
+ * </ul>
+ *
+ * A wrapper must operate on a matching instance type. Specifically: when {@code X} is
+ * <em>any-type</em>, it is not type checked during deployment. Otherwise, it is a deployment
+ * problem if {@code X} is not assignable from the class type of the bean class to which
+ * the invokable method belongs. {@code Y} and {@code Z} are not type checked during deployment.
+ * <p>
+ * When a wrapper is registered, 2 invokers for the same invokable method are created.
+ * The <em>inner</em> invoker applies all lookups and transformations, as described in previous
+ * sections, and invokes the invokable method. The <em>outer</em> invoker calls the wrapper
+ * with the passed instance and arguments and an instance of the inner invoker. The outer invoker
+ * is returned by this invoker builder.
+ * <p>
+ * In other words, the outer invoker is equivalent to the following class:
+ *
+ * <pre>
+ * class InvokerWrapper implements Invoker&lt;X, Z&gt; {
+ *     Z invoke(X instance, Object[] arguments) {
+ *         // obtain the invoker as if no wrapper existed
+ *         Invoker&lt;X, Y&gt; invoker = obtainInvoker();
+ *         return SomeClass.wrap(instance, arguments, invoker);
+ *     }
+ * }
+ * </pre>
+ *
+ * If the wrapper returns normally, the outer invoker returns its return value, unless the wrapper
+ * is declared {@code void}, in which case the outer invoker returns {@code null}. If the wrapper
+ * throws an exception, the outer invoker rethrows it directly.
+ * <p>
+ * The wrapper is supposed to call the invoker it is passed, but does not necessarily have to.
+ * The wrapper may call the invoker multiple times. The wrapper must not use the invoker
+ * in any other way; specifically, it is forbidden to store the invoker instance anywhere
+ * or pass it to other methods that do not follow these rules. Doing so leads to non-portable
+ * behavior.
+ *
+ * <h2>Type checking</h2>
+ *
+ * An invoker created by this builder has relaxed type checking rules, when compared to
+ * the description in {@link Invoker#invoke(Object, Object[]) Invoker.invoke()}, depending
+ * on configured lookups, transformers and wrapper. Some types are checked during
+ * deployment, as described in previous sections. Other types are checked during invocation,
+ * at the very least due to the type checks performed implicitly by the JVM. The lookups,
+ * transformers and the wrapper must arrange the inputs and outputs so that when the invokable
+ * method is eventually invoked, the rules described in
+ * {@link Invoker#invoke(Object, Object[]) Invoker.invoke()} all hold.
+ * <p>
+ * TODO specify what happens when a transformer/wrapper declares a parameter of a primitive type
+ *  but the actual value passed to the invoker is `null` (the transformer should get a zero value?)
```

Comment:
I'd expect `NullPointerException` in this case, as you would get if you tried to unbox a `null`.